src/Agent.Worker/ExecutionContext.cs

```diff
@@ -244,23 +244,30 @@ public void InitializeJob(JobRequestMessage message)
 
             // Initialize the environment.
             Endpoints = message.Environment.Endpoints;
-            Variables = new Variables(HostContext, message.Environment.Variables);
+            List<string> warnings;
+            Variables = new Variables(HostContext, message.Environment.Variables, out warnings);
```

Comment:
Recursive expansion for the variables happens during construction (called from the job-level execution context's InitializeJob method, which is in turn called from jobrunner).